File: Week_04/32.最长有效括号.java
/*
 * @lc app=leetcode.cn id=32 lang=java
 *
 * [32] 最长有效括号
 */

// @lc code=start
class Solution {
    public int longestValidParentheses(String s) {
        int n = s.length(), res = 0;
        // 状态数组：dp[i], 
        // dp equation: '(': 0, ')': i-1:'('; i-1:')';
        int[] dp = new int [n];
        if (n == 0 || s == null) return res;
        dp[0] = 0;
        for (int i = 1; i < n; i++) {
            if (s.charAt(i) == ')') {
                if (s.charAt(i-1) == '(') {
                    dp[i] = (i >= 2? dp[i-2]: 0) + 2;
                }
                else if (i - dp[i-1] > 0 && s.charAt(i - dp[i-1] - 1) == '(') {
                    // dp[i] = 自身构成的单一括号对 + 单一括号对之内的元素的个数 + 匹配的左括号的左边的有效括号长度
                    dp[i] = 2 + dp[i - 1] + (i - dp[i-1] >= 2? dp[i - dp[i - 1] - 2]: 0);
                }
                res = Math.max(res, dp[i]);
            }
        }
        return res;

    }
}
// @lc code=end

File: Week_06/32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python3
#
# [32] 最长有效括号
#

# @lc code=start
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        # subproblem: 
        # dp[n]
        # 2
        
        dp = [0] * len(s)
        res = 0
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    if i == 1:
                        dp[i] = 2
                    else:
                        dp[i] = dp[i-2] + 2
                elif i >= dp[i-1] + 1 and s[i - dp[i-1] - 1] == '(':
                    dp[i] = dp[i-dp[i-1]-2] + dp[i-1] + 2 if i > dp[i-1] + 1 else dp[i-1] + 2
            res = max(res, dp[i])
        return res
        
# @lc code=end

